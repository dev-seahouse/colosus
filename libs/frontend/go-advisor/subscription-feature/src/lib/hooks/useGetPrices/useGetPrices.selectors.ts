import { useQuery } from '@tanstack/react-query';

import { useCallback } from 'react';

import { getPricesQuery } from './useGetPrices';
import type { GetPricesData } from './useGetPrices';

/**
 * hoook to return available prices
 */
export const useSelectPricesQuery = () => {
  return useQuery({
    ...getPricesQuery(),
    select: useCallback((data: GetPricesData) => {
      if (!data || !data.data) {
        return [];
      }

      return data.data ?? [];
    }, []),
  });
};
