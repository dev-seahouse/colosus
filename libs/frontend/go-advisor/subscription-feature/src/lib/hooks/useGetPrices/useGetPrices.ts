import { useQuery } from '@tanstack/react-query';
import { StripeIntegrationPricesApi } from '@bambu/api-client';
import type { StripeIntegrationGetPricesResponseDto } from '@bambu/api-client';
import type { QueryClient } from '@tanstack/react-query';

export type GetPricesData = StripeIntegrationGetPricesResponseDto;

export const getPricesQuery = () => ({
  queryKey: ['getPrices'],
  queryFn: async () => {
    const stripeIntegrationPricesApi = new StripeIntegrationPricesApi();
    const res = await stripeIntegrationPricesApi.getPrices();

    return res.data;
  },
});

export const getPricesLoader =
  (queryClient: QueryClient) => async (): Promise<GetPricesData> => {
    const query = getPricesQuery();

    // ⬇️ return data or fetch it
    return (
      queryClient.getQueryData(query.queryKey) ??
      (await queryClient.fetchQuery(query))
    );
  };

export interface UseGetPricesOptions {
  initialData?: Partial<GetPricesData>;
}
export function useGetPrices({ initialData }: UseGetPricesOptions = {}) {
  return useQuery({
    ...getPricesQuery(),
    initialData,
  });
}

export default useGetPrices;
