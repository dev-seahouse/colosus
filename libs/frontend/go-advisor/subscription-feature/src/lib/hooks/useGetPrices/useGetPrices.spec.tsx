import {
  renderHook,
  queryClientWrapper,
  waitFor,
} from '@bambu/react-test-utils';

import useGetPrices from './useGetPrices';

describe('useGetPrices', () => {
  it('should render successfully', async () => {
    const { result } = renderHook(() => useGetPrices(), {
      wrapper: queryClientWrapper,
    });

    await waitFor(() => expect(result.current.isSuccess).toBe(true));

    expect(result.current.data).toMatchSnapshot();
  });
});
