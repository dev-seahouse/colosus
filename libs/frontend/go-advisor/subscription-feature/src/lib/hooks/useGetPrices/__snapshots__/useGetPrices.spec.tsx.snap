// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useGetPrices > should render successfully 1`] = `
{
  "data": [
    {
      "active": true,
      "billing_scheme": "per_unit",
      "created": 1680101258,
      "currency": "usd",
      "custom_unit_amount": null,
      "id": "price_1Mr0HmFabpOjRn7HwuvraNAD",
      "livemode": false,
      "lookup_key": null,
      "metadata": {},
      "nickname": null,
      "object": "price",
      "product": {
        "active": false,
        "attributes": [],
        "created": 1680101258,
        "default_price": null,
        "description": "(created by Stripe CLI)",
        "id": "prod_NcEl4gKh6VbQdV",
        "images": [],
        "livemode": false,
        "metadata": {},
        "name": "myproduct",
        "object": "product",
        "package_dimensions": null,
        "shippable": null,
        "statement_descriptor": null,
        "tax_code": null,
        "type": "service",
        "unit_label": null,
        "updated": 1680101661,
        "url": null,
      },
      "recurring": null,
      "tax_behavior": "unspecified",
      "tiers_mode": null,
      "transform_quantity": null,
      "type": "one_time",
      "unit_amount": 1500,
      "unit_amount_decimal": "1500",
    },
    {
      "active": true,
      "billing_scheme": "per_unit",
      "created": 1679897237,
      "currency": "myr",
      "custom_unit_amount": null,
      "id": "price_1Mq9D7FabpOjRn7HWILE9Cro",
      "livemode": false,
      "lookup_key": null,
      "metadata": {},
      "nickname": "Connect",
      "object": "price",
      "product": {
        "active": true,
        "attributes": [],
        "created": 1679897236,
        "default_price": "price_1Mq9D7FabpOjRn7HWILE9Cro",
        "description": "Product used for POC",
        "id": "prod_NbLut6uJw85qmi",
        "images": [],
        "livemode": false,
        "metadata": {
          "lookup_key": "TESTING",
        },
        "name": "Ben POC Product",
        "object": "product",
        "package_dimensions": null,
        "shippable": null,
        "statement_descriptor": null,
        "tax_code": null,
        "type": "service",
        "unit_label": null,
        "updated": 1679905720,
        "url": null,
      },
      "recurring": {
        "aggregate_usage": null,
        "interval": "month",
        "interval_count": 1,
        "trial_period_days": null,
        "usage_type": "licensed",
      },
      "tax_behavior": "unspecified",
      "tiers_mode": null,
      "transform_quantity": null,
      "type": "recurring",
      "unit_amount": 2000,
      "unit_amount_decimal": "2000",
    },
  ],
}
`;
